data/code/lib.c

```diff
@@ -589,3 +593,793 @@ void check_lap_ex()
 #endif
 }
 #endif
+
+// This is in its own section despite not needing to be for now as it will have
+// to handle cases where one custom material is enabled and another is not.
+#if EXTENDED_TERRAIN_TYPES
+#if !GM4E01_DEBUG_BUILD
+// Check against all of the custom material flags enabled by the patch.
+int is_terrain_type_mod(char* terrain_type)
+{
+    int ret = 0;
+    if (*terrain_type == 0xB0)
+    {
+        ret = 1;
+    }
+    return ret;
+}
+
+// This and skip_item are separated to make bug reports easier to diagnose in the future.
+int should_return_fake_code(char* const ground)
+{
+    int* ground_materials = (int*)(ground + 0x20);  // Pointer to the triangle's material in memory.
+    int ret = 0;
+
+    if (*ground_materials != 0)
```

Comment:
This is a pointer to a specific triangle's data structure, seemingly the one a Kart is currently calculated to be interacting with. It's the exact data structure documented here:
- https://mkdd.org/wiki/BCO_(File_Format)#Triangles

It's important to note that this seems to be loaded even while a Kart is not actually colliding with it; it is a pointer to whatever triangle is directly underneath (?) the Kart. It's probable you know much more about this than I do.

The way I understand it, a more accurate name for it may be something like `collision_triangle_underneath`.

Reply:
Since we know it's pointing to that structure, how about actually declaring the structure?

```cpp
typedef struct
{
    int point_indexes[3];
    float unknown;
    short normal[3];
    char terrain_type;
    char terrain_type_part_two;
    // ...
} CollisionTriangle;

int should_return_fake_code(const char* const ground)
{
    const int* const ground_materials =
        (const int*)(ground + 0x20);  // Pointer to the triangle's material in memory.

    const CollisionTriangle* const triangle = (const CollisionTriangle*)*ground_materials;
    if (triangle)
    {
        return is_extended_terrain_type(triangle->terrain_type);
    }

    return 0;
}
```

Similarly, it seems we know that `ground` is pointing to a `CrsGround` structure. We could declare that structure too (or reduced version of it that includes the members we are interested in).

Reply:
If you don't mind me asking, what is the purpose of defining the Collision Triangle in this way - is it simply just for readability? It definitely helps that `CollisionTriangle` is already well documented.
For defining a concatenated `CrsGround` struct with just the parts we need, we would just assign a buffer and then the relevant members at the correct offsets? I don't love that it's so incomplete, but I would also prefer not to reverse engineer the entire `CrsGround` structure.

Reply:
I see what you mean... The few known offsets in `KartBody` leave quite a few unknown regions... However, I still think it's worth it. So many of those `*(type*)(address + offset)` expressions could go away nicely.

Reply:
The majority of the time I spent implementing <KartBody> as a struct, which has been about 20~ hours over the past two months, was spent just trying to get it to work. I definitely overcomplicated it.

On that note, I did not choose to make a struct for the RaceManager object as it would require three structs as it's just three pointers in a row, used in this patch just for getting the single bit that represents Mirror Mode.

Reply:
After helping other people, I decided to just do the RaceManager stuff as structs as well, so there are no more of those `*(type*)(address + offset)` expressions aside from at the very start of the hook function.